datafusion/optimizer/src/rewrite.rs

```diff
@@ -0,0 +1,199 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Trait to make LogicalPlan rewritable
+
+use datafusion_common::Result;
+
+use datafusion_expr::LogicalPlan;
+
+/// a Trait for marking tree node types that are rewritable
+pub trait TreeNodeRewritable: Clone {
```

Comment:
This appears to largely be the same as https://github.com/apache/arrow-datafusion/blob/146a949218ec970784974137277cde3b4e547d0a/datafusion/physical-expr/src/rewrite.rs#L25 but they are both generic, 

Would it be possible to move the trait into `datafusion_common` and only keep the impls in logical and physical plan?

Reply:
In the past, I had tried to move the Trait to a common place but failed due to Rust's orphan rule limitations.
Either the trait or the type for which we are implementing the trait must be defined in the same crate. The major problem is the `Arc` struct is from the std crate.

`impl TreeNodeRewritable for Arc<dyn ExecutionPlan>`

`impl TreeNodeRewritable for Arc<dyn PhysicalExpr>`